feature: add optional support for moved in / out blips

setchy opened the pull request:
Hot on the heels of the [v1.0.0](https://github.com/thoughtworks/build-your-own-radar/releases/tag/v1.0.0) launch this week, I thought I'd see whether I could self-serve #177 

Adds optional support for moved in and out blip drawing.

Below are two datasets for Thoughtworks Volume 27 sample (csv) that can be used to demonstrate/test:
* [Thoughtworks Volume 27 with movements](https://raw.githubusercontent.com/setchy/thoughtworks-tech-radar-volumes/main/volumes/Thoughtworks%20Technology%20Radar%20Volume%2027.csv)
* [Thoughtworks Volume 27 without movements](https://docs.google.com/spreadsheets/d/1sL6j3eYxsiXMtgELi52pn6XMzk_L4ZtfxYr1KjNYZpo/edit#gid=0)


![Screenshot 2023-04-22 at 4 18 08 PM](https://user-images.githubusercontent.com/386277/233804679-65e2d974-66db-4a51-b562-180d35aa03dc.png)

setchy commented:
@devansh-sharma-tw - appreciate your feedback re this feature proposal

setchy commented:
@marisahoenig - would you be able take this to the team for review please

devansh-sharma-tw commented:
Thanks for raising this @setchy ! We're internally discussing on adding the moved in/out status as well (to keep BYOR inline with our Thoughtworks Radar).
This PR helps a lot for this and we'll update here soon regarding the feature.
Thanks!

setchy commented:
@devansh-sharma-tw - hope all is well. Just checking in to see if your team has had a chance to review and consider this PR.

marisahoenig commented on spec/util/inputSanitizer-spec.js:
```diff
@@ -107,6 +107,8 @@ describe('Input Santizer for Protected sheet', function () {
       ring: '',
       quadrant: '',
       isNew: '',
+      hasMovedIn: '',
```
Thinking about this more, I like the idea of having one field that can be utilized instead of `isNew` since a blip can be new OR moved in OR moved out OR a refresh.

setchy replied:
I agree, a new field `status` or similar  

Would you be willing to make this a breaking change, or want to support backwards compatibility with datasets that are using `isNew`? 

Can you elaborate on your definition of `refresh`.  Is this for blips that are removed from the radar and re-appear, or something different?

marisahoenig replied:
I'd prefer making it backward compatible. So, users should have the option to _either_ use `isNew` or `status`. If there's a value in `status`, we would use that field.

And here's where the differences between TW Tech Radar and BYOR come back 😊 The "refresh" I'm referring to is what shows as "No change" on https://www.thoughtworks.com/radar. Basically, the blip has appeared before and we're blipping it again, but it stays in the same ring as last time. In BYOR, this would be a `false` for `isNew`.

setchy replied:
Ok great, that clears things up.  Would you want to keep it as `refresh`, or prefer something different?

marisahoenig replied:
Since `refresh` isn't too intuitive, I'm open to suggestions! I was thinking maybe `existing` but not sure on that. `notNew` seems not great too.

setchy replied:
I think either `existing` or `noChange` would work. Whichever we use should match the terminology used in the legend.

marisahoenig reviewed (changes requested):
Adding some suggestions. I think it's worth adding one new field and letting users have the option to use it OR the `isNew` field. Would love your thoughts @setchy

setchy commented on src/graphing/blips.js:
```diff
@@ -104,9 +104,20 @@ function findBlipCoordinates(blip, minRadius, maxRadius, startAngle, allBlipCoor
   return blip.isGroup()
     ? `\`${blip.ring().name()} ring, group of ${blip.blipText()}`
-    : `${blip.ring().name()} ring, ${blip.name()}, ${blip.isNew() ? 'new' : 'existing'} blip.`
+    : `${blip.ring().name()} ring, ${blip.name()}, ${blipType} blip.`
```
Agree :)

setchy commented on src/stylesheets/_quadrants.scss:
```diff
@@ -402,8 +402,8 @@
 
-    img:nth-child(2) {
-      margin-left: 48px;
+    img:nth-child(n + 2) {
```
I don't off the top of my head, it's been too long...

marisahoenig replied:
Fair! All good then.

setchy replied:
Ah, i think I recall, it was to support the legend going from 2 items to 3 items

setchy commented:
@marisahoenig - I have updated this PR to support a new, optional column/field called `status`.  

Accepted values are case-insensitive `new`, `no change`, `moved in`, `moved out`.  I welcome any feedback on alternate option names (eg: `existing` or `unchanged` etc)

To demonstrate, I've updated my [TW volume datasets](https://github.com/setchy/thoughtworks-tech-radar-volumes/tree/main/volumes), eg [TW Volume 28](https://github.com/setchy/thoughtworks-tech-radar-volumes/blob/main/volumes/Thoughtworks%20Technology%20Radar%20Volume%2028.csv).

All existing datasets using `isNew` should continue to work.

setchy commented:
> [!NOTE]
> in the interim I've deployed my BYOR enhancements @ https://radar.setchy.io, which bring it up to the functionality found on https://thoughtworks.com/radar

danielkoch commented:
@marisahoenig: Any plans or update on this one?

marisahoenig commented on spec/graphing/blips-spec.js:
```diff
@@ -187,12 +187,89 @@ describe('Blips', function () {
+          name: () => 'ring1',
+        }
+      },
+      blipText: () => '12 New Blips',
```
For this test, it should be existing blips — also the `actual` statement should be something like "group of 12 Existing Blips"

marisahoenig replied:
Should there be a status() value here?

setchy replied:
Since this test case wasn't for grouped blips, I've removed it

marisahoenig commented on spec/graphing/blips-spec.js:
```diff
@@ -187,12 +187,89 @@ describe('Blips', function () {
+    }
+
+    const actual = blipAssistiveText(blip)
+    expect(actual).toEqual('ring1 ring, blip1, moved in blip.')
```
`actual` here needs to be revised too. And I think status() should be added as well?

setchy replied:
Ah yes, great catch.  Upon inspecting thoughtworks.com/radar I can see that the assistive text format has been updated/fixed in the past 12mths.  

I'll pushed an update which aligns this with the following observed format
`<ring name> ring, <blip name>, <New|Moved in|Moved out|No change>.`

marisahoenig commented on spec/graphing/blips-spec.js:
```diff
@@ -187,12 +187,89 @@ describe('Blips', function () {
+    }
+
+    const actual = blipAssistiveText(blip)
+    expect(actual).toEqual('ring1 ring, blip1, moved out blip.')
```
Same comment here

setchy replied:
Since this test case wasn't for grouped blips, I've removed it

marisahoenig commented on src/graphing/radar.js:
```diff
@@ -828,12 +828,30 @@ const Radar = function (size, radar) {
+
+      for (var blip of quadrant.blips()) {
+        if (blip.status() !== '') {
+          hasMovements = true
```
Could we just `return true` here instead (and then `return false` instead of `return hasMovements`) to get rid of the `hasMovements` variable?

setchy replied:
Deal 🤝

marisahoenig commented on src/graphing/radar.js:
```diff
@@ -828,12 +828,30 @@ const Radar = function (size, radar) {
     if (featureToggles.UIRefresh2022) {
-      renderRadarLegends(radarElement)
+      let hasMovements = hasMovementData(quadrants)
+      renderRadarLegends(radarElement, hasMovements)
```
Consolidate into:
`renderRadarLegends(radarElement, hasMovementData(quadrants))` since we don't use the `hasMovements` var again

setchy replied:
Deal 🤝

marisahoenig commented on spec/models/blip-spec.js:
```diff
@@ -67,6 +67,30 @@ describe('Blip', function () {
     expect(blip.isNew()).toBe(false)
   })
 
+  it('status is new', function () {
```
Can you add topic and description to these tests so it's easier to follow which params are which?

setchy replied:
Sorted.

marisahoenig commented on src/graphing/blips.js:
```diff
@@ -104,9 +104,21 @@ function findBlipCoordinates(blip, minRadius, maxRadius, startAngle, allBlipCoor
+    blipType = 'moved in'
+  } else if (blip.hasMovedOut()) {
+    blipType = 'moved out'
+  } else {
```
Should the `hasNoChange` fxn be incorporated here?

setchy replied:
Simplified to use blip.status directly (previously the assistive text was formatted in a way that wasn't compatible.  Now, not the case :)

marisahoenig commented on src/graphing/blips.js:
```diff
@@ -104,9 +104,21 @@ function findBlipCoordinates(blip, minRadius, maxRadius, startAngle, allBlipCoor
   return blip.isGroup()
     ? `\`${blip.ring().name()} ring, group of ${blip.blipText()}`
-    : `${blip.ring().name()} ring, ${blip.name()}, ${blip.isNew() ? 'new' : 'existing'} blip.`
+    : `${blip.ring().name()} ring, ${blip.name()}, ${blipType} blip.`
```
Very nice!

marisahoenig reviewed (changes requested):
Hey @setchy I left a few comments. Please resolve and also make sure the most recent code from the main branch of this repo is loaded in. I'll chat with the team next week to see if we can get this added. The slowdown was related to the most recent Radar release, and we need to make sure changes like this are reflected on our live site as well. Apologies for the delay — it's a great add to the tool.

setchy commented:
> Hey @setchy I left a few comments. Please resolve and also make sure the most recent code from the main branch of this repo is loaded in. I'll chat with the team next week to see if we can get this added. The slowdown was related to the most recent Radar release, and we need to make sure changes like this are reflected on our live site as well. Apologies for the delay — it's a great add to the tool.

Thanks @marisahoenig, appreciate the feedback.  I've gone through and addressed each of them 😄.  

No problems about the delay, too - completely understand.

marisahoenig reviewed (approved):
Thanks for all the changes, @setchy! As mentioned, I'll chat with the team next week about getting this incorporated and seeing if there is anything else that needs to be updated on the team's end.

setchy commented:
@will-amaral - appreciate your thoughts on this PR
